File: iOS/Crisis Connect/Features/SOS/SosCloudReporter.swift
//
//  SosCloudReporter.swift
//  Crisis Connect
//
//  Victim-side internet uplink for the SOS broadcast — the iOS port of Android's SosCloudReporter.
//
//  The BLE broadcast is the primary, offline-first channel and must never wait on this class. When
//  usable connectivity exists (now or later — store-and-forward), the active SOS is additionally
//  self-reported to the agency dashboard through the `reportSosSignal` callable, refreshed on a
//  battery-aware heartbeat, and finally marked resolved when the victim stops the broadcast. A
//  resolve that cannot be delivered (offline stop) is persisted and retried on the next app
//  foreground. The signalId is the same rescue device id the BLE advert carries, so the dashboard
//  merges the self-report with any rescuer-side sightings of the same victim.
//

import BackgroundTasks
import Combine
import Foundation
import FirebaseAuth
import FirebaseFunctions
import CoreLocation
import Network
import UIKit

final class SosCloudReporter: ObservableObject, @unchecked Sendable {
    static let shared = SosCloudReporter()

    enum State: Equatable {
        case idle
        /// SOS is active but there is no usable network yet; the report is queued locally.
        case waitingForNetwork
        case sending
        case reported(lastReportedAt: Date)
        /// Last attempt over a live network failed; will retry automatically.
        case failed
    }

    @Published private(set) var state: State = .idle

    private static let callableName = "reportSosSignal"
    private static let functionsRegion = "us-central1"
    private static let statusActive = "active"
    private static let statusResolved = "resolved"

    private static let callTimeout: TimeInterval = 20
    private static let heartbeatInterval: TimeInterval = 45
    private static let heartbeatIntervalLowBattery: TimeInterval = 90
    private static let lowBatteryThresholdPercent = 20
    private static let offlinePollInterval: TimeInterval = 10
    private static let failureRetryDelay: TimeInterval = 20
    private static let signInTimeout: TimeInterval = 10

    private enum Keys {
        static let pendingResolveSignalId = "sos.cloudReport.pendingResolveSignalId"
        static let hasReportedActive = "sos.cloudReport.hasReportedActive"
    }

    private let monitor = NWPathMonitor()
    private let monitorQueue = DispatchQueue(label: "sos.cloud.reporter.reachability")
    private let stateLock = NSLock()
    private var isOnline = false
    private var heartbeatTask: Task<Void, Never>?

    private init() {
        monitor.pathUpdateHandler = { [weak self] path in
            guard let self else { return }
            self.stateLock.lock()
            self.isOnline = path.status == .satisfied
            self.stateLock.unlock()
        }
        monitor.start(queue: monitorQueue)
    }

    private var hasUsableInternet: Bool {
        stateLock.lock(); defer { stateLock.unlock() }
        return isOnline
    }

    /// Starts the uplink for a freshly started SOS broadcast. Runs entirely on its own task so a
    /// slow network call can never delay or block BLE startup or teardown.
    func onSosStarted() {
        guard !PlatformRuntime.isRunningTests else { return }
        heartbeatTask?.cancel()
        // A new broadcast supersedes any queued resolve for the same device signal.
        UserDefaults.standard.removeObject(forKey: Keys.pendingResolveSignalId)
        UserDefaults.standard.set(false, forKey: Keys.hasReportedActive)
        // @Published `state` now drives the active-SOS UI, so every mutation hops to the main actor.
        let initialState: State = hasUsableInternet ? .sending : .waitingForNetwork
        Task { @MainActor in self.state = initialState }

        heartbeatTask = Task { [weak self] in
            await MainActor.run { UIDevice.current.isBatteryMonitoringEnabled = true }
            while let self, !Task.isCancelled {
                let sessionActive = await MainActor.run { SOSBroadcastManager.shared.isSessionActive }
                guard sessionActive else { break }

                guard self.hasUsableInternet else {
                    await MainActor.run { self.state = .waitingForNetwork }
                    try? await Task.sleep(nanoseconds: UInt64(Self.offlinePollInterval * 1_000_000_000))
                    continue
                }
                await MainActor.run { self.state = .sending }
                let (location, battery) = await MainActor.run {
                    (
                        SOSSignalLocationCoordinator.shared.currentVictimLocationPayload(),
                        Self.batteryPercent()
                    )
                }
                let reported = await self.sendReport(
                    status: Self.statusActive,
                    location: location,
                    batteryPercent: battery
                )
                if reported {
                    UserDefaults.standard.set(true, forKey: Keys.hasReportedActive)
                    await MainActor.run { self.state = .reported(lastReportedAt: Date()) }
                    let lowBattery = battery.map { $0 < Self.lowBatteryThresholdPercent } ?? false
                    let interval = lowBattery ? Self.heartbeatIntervalLowBattery : Self.heartbeatInterval
                    try? await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
                } else {
                    await MainActor.run { self.state = .failed }
                    try? await Task.sleep(nanoseconds: UInt64(Self.failureRetryDelay * 1_000_000_000))
                }
            }
        }
    }

    /// Marks the broadcast resolved on the dashboard. Best-effort immediately; when offline (or on
    /// failure) the resolve is persisted and retried on the next app foreground. Only signals that
    /// actually reported as active get resolved, so a fully-offline SOS session never writes a
    /// phantom "resolved" marker.
    func onSosStopped() {
        guard !PlatformRuntime.isRunningTests else { return }
        heartbeatTask?.cancel()
        heartbeatTask = nil
        Task { @MainActor in self.state = .idle }
        guard UserDefaults.standard.bool(forKey: Keys.hasReportedActive) else { return }
        let signalId = SecureLocalStore.shared.getOrCreateRescueDeviceId()
        UserDefaults.standard.set(signalId, forKey: Keys.pendingResolveSignalId)
        UserDefaults.standard.set(false, forKey: Keys.hasReportedActive)
        Task { [weak self] in
            guard let self else { return }
            if self.hasUsableInternet,
               await self.sendReport(status: Self.statusResolved, location: nil, batteryPercent: nil) {
                UserDefaults.standard.removeObject(forKey: Keys.pendingResolveSignalId)
            }
            // Otherwise the pending resolve stays persisted; deliverPendingResolve retries it on the
            // next foreground, and a network-gated background task narrows the offline-stop gap.
            SosFollowUpBackgroundManager.shared.scheduleIfNeeded()
        }
    }

    /// True when an offline stop left a "resolved" report still awaiting delivery — the trigger the
    /// background follow-up task checks before scheduling itself.
    static var hasPendingResolve: Bool {
        UserDefaults.standard.string(forKey: Keys.pendingResolveSignalId) != nil
    }

    /// Attempts delivery of a persisted resolve (called on app foreground — the iOS stand-in for
    /// Android's SosReportWorker).
    func deliverPendingResolve() async {
        guard let pending = UserDefaults.standard.string(forKey: Keys.pendingResolveSignalId) else {
            return
        }
        // The victim restarted SOS meanwhile; the active heartbeat owns the signal again.
        let sessionActive = await MainActor.run { SOSBroadcastManager.shared.isSessionActive }
        if sessionActive { return }
        guard hasUsableInternet else { return }
        if await sendReport(
            status: Self.statusResolved,
            location: nil,
            batteryPercent: nil,
            signalIdOverride: pending
        ) {
            UserDefaults.standard.removeObject(forKey: Keys.pendingResolveSignalId)
        }
    }

    // MARK: - Report delivery

    /// Returns true when the report was accepted by the backend. Payload fields mirror Android's
    /// SosCloudReporter exactly (the dashboard consumes one schema for both platforms).
    private func sendReport(
        status: String,
        location: SOSSignalLocationPayload.GPSPayload?,
        batteryPercent: Int?,
        signalIdOverride: String? = nil
    ) async -> Bool {
        if status == Self.statusActive && location == nil {
            NSLog("SosCloudReporter: skipping report; no usable location fix yet")
            return false
        }
        guard await ensureSignedIn() else {
            NSLog("SosCloudReporter: skipping report; no Firebase identity available")
            return false
        }
        let signalId = signalIdOverride ?? SecureLocalStore.shared.getOrCreateRescueDeviceId()
        var payload: [String: Any] = [
            "signalId": signalId,
            "status": status
        ]
        if let location {
            payload["lat"] = location.latitude
            payload["lon"] = location.longitude
            payload["capturedAtMillis"] = location.capturedAtMillis
            payload["provider"] = location.provider
            if let accuracy = location.horizontalAccuracyMeters {
                payload["accuracyMeters"] = accuracy
            }
        }
        if let batteryPercent {
            payload["batteryPercent"] = batteryPercent
        }
        if let countryIso = await resolveCountryIso(location: location) {
            payload["countryIso"] = countryIso
        }

        let call = Task {
            try await Functions.functions(region: Self.functionsRegion)
                .httpsCallable(Self.callableName)
                .callAsync(payload)
        }
        let timeout = Task {
            try? await Task.sleep(nanoseconds: UInt64(Self.callTimeout * 1_000_000_000))
            call.cancel()
        }
        defer { timeout.cancel() }
        do {
            _ = try await call.value
            AppAnalytics.sosCloudReported(status: status)
            return true
        } catch {
            NSLog("SosCloudReporter: report failed status=%@: %@", status, String(describing: error))
            return false
        }
    }

    /// Restores the Firebase session or signs in anonymously, time-bounded.
    private func ensureSignedIn() async -> Bool {
        if Auth.auth().currentUser != nil { return true }
        let signIn = Task { try? await Auth.auth().signInAnonymously() }
        let timeout = Task {
            try? await Task.sleep(nanoseconds: UInt64(Self.signInTimeout * 1_000_000_000))
            signIn.cancel()
        }
        defer { timeout.cancel() }
        return await signIn.value != nil
    }

    /// Best-effort ISO country for agency routing: reverse-geocode the fix, else the device
    /// locale's region. Any failure just omits the hint (the dashboard routes to International).
    private func resolveCountryIso(location: SOSSignalLocationPayload.GPSPayload?) async -> String? {
        if let location {
            let clLocation = CLLocation(latitude: location.latitude, longitude: location.longitude)
            let geocoded = try? await CLGeocoder().reverseGeocodeLocation(clLocation)
            if let iso = geocoded?.first?.isoCountryCode?.trimmingCharacters(in: .whitespacesAndNewlines),
               !iso.isEmpty {
                return iso.uppercased()
            }
        }
        return Locale.current.region?.identifier.uppercased()
    }

    /// Roaming-safe ISO country for the emergency-call card: reverse-geocode the victim's live GPS
    /// fix using the same logic the dashboard report uses. Returns nil when there is no usable fix,
    /// so the caller keeps its device-locale default. Never throws.
    func currentGeocodedCountryIso() async -> String? {
        let location = await MainActor.run {
            SOSSignalLocationCoordinator.shared.currentVictimLocationPayload()
        }
        guard let location else { return nil }
        return await resolveCountryIso(location: location)
    }

    @MainActor
    private static func batteryPercent() -> Int? {
        let level = UIDevice.current.batteryLevel
        guard level >= 0 else { return nil }
        return Int((level * 100).rounded())
    }
}

/// Store-and-forward tail of an SOS session — the iOS stand-in for Android's SosReportWorker
/// (WorkManager, NetworkType.CONNECTED). When the victim stops the broadcast offline, the dashboard
/// "resolved" report and the contacts' all-clear are persisted; this network-gated background task
/// delivers whatever is left even if the app is never reopened. iOS background execution is
/// best-effort, so the app-foreground hooks remain the primary retry path — this only narrows the
/// gap. A no-op whenever nothing is pending. Registration mirrors CertificateRenewalBackgroundManager.
final class SosFollowUpBackgroundManager {
    static let shared = SosFollowUpBackgroundManager()
    static let taskIdentifier = "com.auralis.crisisconnect.sos-followup"

    private static let retryDelay: TimeInterval = 60

    private var hasRegistered = false

    private init() {}

    func register() {
        guard !PlatformRuntime.isRunningTests else { return }
        guard !hasRegistered else { return }
        hasRegistered = BGTaskScheduler.shared.register(
            forTaskWithIdentifier: Self.taskIdentifier,
            using: nil
        ) { task in
            guard let processingTask = task as? BGProcessingTask else {
                task.setTaskCompleted(success: false)
                return
            }
            Self.shared.handle(processingTask)
        }
        if !hasRegistered {
            NSLog("Failed to register SOS follow-up background task.")
        }
    }

    /// Arms the network-gated task when an offline stop left a resolve or all-clear undelivered;
    /// cancels the request when nothing is pending. Safe to call repeatedly (cancel-then-submit).
    func scheduleIfNeeded() {
        guard !PlatformRuntime.isRunningTests else { return }
        guard SosCloudReporter.hasPendingResolve || SosContactNotifier.hasPendingAllClear else {
            BGTaskScheduler.shared.cancel(taskRequestWithIdentifier: Self.taskIdentifier)
            return
        }

        BGTaskScheduler.shared.cancel(taskRequestWithIdentifier: Self.taskIdentifier)
        let request = BGProcessingTaskRequest(identifier: Self.taskIdentifier)
        request.requiresNetworkConnectivity = true
        request.requiresExternalPower = false
        request.earliestBeginDate = Date(timeIntervalSinceNow: Self.retryDelay)

        do {
            try BGTaskScheduler.shared.submit(request)
        } catch {
            NSLog("Failed to schedule SOS follow-up background task: %@", String(describing: error))
        }
    }

    private func handle(_ task: BGProcessingTask) {
        let worker = Task(priority: .background) {
            await SosCloudReporter.shared.deliverPendingResolve()
            _ = await SosContactNotifier.shared.deliverPendingAllClear()
        }
        task.expirationHandler = {
            worker.cancel()
        }

        Task {
            await worker.value
            task.setTaskCompleted(success: true)
            // Anything still undelivered (e.g. the network dropped mid-run) reschedules itself.
            Self.shared.scheduleIfNeeded()
        }
    }
}

File: Android/app/src/main/java/com/auralis/crisisconnect/service/sos/SosCloudReporter.kt
package com.auralis.crisisconnect.service.sos

import android.content.Context
import android.content.SharedPreferences
import android.location.Geocoder
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.telephony.TelephonyManager
import android.util.Log
import com.auralis.crisisconnect.analytics.Analytics
import com.auralis.crisisconnect.data.database.LocalKeyStorage
import com.auralis.crisisconnect.service.BlePeerIdentityUtils
import com.auralis.crisisconnect.service.GattSOSServerService
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.functions.FirebaseFunctions
import java.util.Locale
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull

/**
 * Victim-side internet uplink for the SOS broadcast.
 *
 * The BLE broadcast is the primary, offline-first channel and must never wait on this class. When
 * usable connectivity exists (now or later — store-and-forward), the active SOS is additionally
 * self-reported to the agency dashboard through the `reportSosSignal` callable, then refreshed on
 * a battery-aware heartbeat, and finally marked resolved when the victim stops the broadcast.
 * A resolve that cannot be delivered (offline stop) is persisted and retried by
 * [SosReportWorker] once a network appears.
 */
object SosCloudReporter {

    sealed interface State {
        data object Idle : State

        /** SOS is active but there is no usable network yet; the report is queued locally. */
        data object WaitingForNetwork : State

        data object Sending : State

        data class Reported(
            /** Human-readable place of the reported fix (e.g. district/city), when resolvable. */
            val placeName: String?,
            val lastReportedAtMillis: Long
        ) : State

        /** Last attempt over a live network failed; will retry automatically. */
        data object Failed : State
    }

    private const val TAG = "SosCloudReporter"
    private const val FUNCTIONS_REGION = "us-central1"
    private const val CALLABLE_NAME = "reportSosSignal"
    private const val PREFS_NAME = "sos_cloud_reporter"
    private const val KEY_PENDING_RESOLVE_SIGNAL_ID = "pending_resolve_signal_id"
    private const val KEY_HAS_REPORTED_ACTIVE = "has_reported_active"

    private const val CALL_TIMEOUT_MS = 20_000L
    private const val HEARTBEAT_INTERVAL_MS = 45_000L
    private const val HEARTBEAT_INTERVAL_LOW_BATTERY_MS = 90_000L
    private const val LOW_BATTERY_THRESHOLD_PERCENT = 20
    private const val OFFLINE_POLL_INTERVAL_MS = 10_000L
    private const val FAILURE_RETRY_DELAY_MS = 20_000L
    private const val SIGN_IN_TIMEOUT_MS = 10_000L

    private val reporterScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var heartbeatJob: Job? = null

    private val _state = MutableStateFlow<State>(State.Idle)
    val state: StateFlow<State> = _state.asStateFlow()

    /**
     * Starts the uplink for a freshly started SOS broadcast. Runs entirely on the reporter's own
     * scope so a slow network call can never delay or block BLE startup or service teardown.
     */
    fun onSosStarted(
        context: Context,
        locationProvider: suspend () -> BlePeerIdentityUtils.PeerLocationSnapshot?,
        batteryProvider: () -> Int?
    ) {
        val appContext = context.applicationContext
        heartbeatJob?.cancel()
        cachedPlace = null
        // A new broadcast supersedes any queued resolve for the same device signal.
        prefs(appContext).edit()
            .remove(KEY_PENDING_RESOLVE_SIGNAL_ID)
            .putBoolean(KEY_HAS_REPORTED_ACTIVE, false)
            .apply()
        _state.value =
            if (appContext.hasUsableInternet()) State.Sending else State.WaitingForNetwork
        heartbeatJob = reporterScope.launch {
            while (isActive && GattSOSServerService.isDeclared.value) {
                if (!appContext.hasUsableInternet()) {
                    _state.value = State.WaitingForNetwork
                    delay(OFFLINE_POLL_INTERVAL_MS)
                    continue
                }
                _state.value = State.Sending
                val battery = runCatching(batteryProvider).getOrNull()
                val location = runCatching { locationProvider() }.getOrNull()
                val reported = sendReport(
                    context = appContext,
                    status = STATUS_ACTIVE,
                    location = location,
                    batteryPercent = battery
                )
                if (reported) {
                    prefs(appContext).edit().putBoolean(KEY_HAS_REPORTED_ACTIVE, true).apply()
                    _state.value = State.Reported(
                        placeName = resolvePlaceName(appContext, location),
                        lastReportedAtMillis = System.currentTimeMillis()
                    )
                    val interval = if (battery != null && battery < LOW_BATTERY_THRESHOLD_PERCENT) {
                        HEARTBEAT_INTERVAL_LOW_BATTERY_MS
                    } else {
                        HEARTBEAT_INTERVAL_MS
                    }
                    delay(interval)
                } else {
                    _state.value = State.Failed
                    delay(FAILURE_RETRY_DELAY_MS)
                }
            }
        }
    }

    /**
     * Marks the broadcast resolved on the dashboard. Best-effort immediately; when offline (or on
     * failure) the resolve is persisted and handed to [SosReportWorker] for delivery once a
     * network appears. Only signals that were actually reported as active get resolved, so a
     * fully-offline SOS session never writes a phantom "resolved" marker.
     */
    fun onSosStopped(context: Context) {
        val appContext = context.applicationContext
        heartbeatJob?.cancel()
        heartbeatJob = null
        _state.value = State.Idle
        if (!prefs(appContext).getBoolean(KEY_HAS_REPORTED_ACTIVE, false)) {
            return
        }
        val signalId = LocalKeyStorage.getOrCreateRescueDeviceId(appContext)
        prefs(appContext).edit()
            .putString(KEY_PENDING_RESOLVE_SIGNAL_ID, signalId)
            .putBoolean(KEY_HAS_REPORTED_ACTIVE, false)
            .apply()
        reporterScope.launch {
            val resolved = appContext.hasUsableInternet() && sendReport(
                context = appContext,
                status = STATUS_RESOLVED,
                location = null,
                batteryPercent = null
            )
            if (resolved) {
                clearPendingResolve(appContext)
            } else {
                SosReportWorker.enqueue(appContext)
            }
        }
    }

    /** Attempts delivery of a persisted resolve; used by [SosReportWorker]. */
    internal suspend fun deliverPendingResolve(context: Context): Boolean {
        val appContext = context.applicationContext
        val pending = prefs(appContext).getString(KEY_PENDING_RESOLVE_SIGNAL_ID, null)
            ?: return true
        // The victim restarted SOS meanwhile; the active heartbeat owns the signal again.
        if (GattSOSServerService.isDeclared.value) {
            return true
        }
        val resolved = sendReport(
            context = appContext,
            status = STATUS_RESOLVED,
            location = null,
            batteryPercent = null,
            signalIdOverride = pending
        )
        if (resolved) {
            clearPendingResolve(appContext)
        }
        return resolved
    }

    internal fun clearPendingResolve(context: Context) {
        prefs(context.applicationContext).edit().remove(KEY_PENDING_RESOLVE_SIGNAL_ID).apply()
    }

    /** Returns true when the report was accepted by the backend. */
    private suspend fun sendReport(
        context: Context,
        status: String,
        location: BlePeerIdentityUtils.PeerLocationSnapshot?,
        batteryPercent: Int?,
        signalIdOverride: String? = null
    ): Boolean {
        if (status == STATUS_ACTIVE && location == null) {
            Log.w(TAG, "Skipping SOS cloud report: no usable location fix yet")
            return false
        }
        val uid = ensureSignedInUid() ?: run {
            Log.w(TAG, "Skipping SOS cloud report: no Firebase identity available")
            return false
        }
        val signalId = signalIdOverride ?: LocalKeyStorage.getOrCreateRescueDeviceId(context)
        val payload = buildMap<String, Any> {
            put("signalId", signalId)
            put("status", status)
            location?.let { fix ->
                put("lat", fix.latitude)
                put("lon", fix.longitude)
                put("capturedAtMillis", fix.capturedAtMillis)
                put("provider", fix.source)
                fix.accuracyMeters?.let { put("accuracyMeters", it.toDouble()) }
            }
            batteryPercent?.let { put("batteryPercent", it) }
            resolveCountryIso(context, location)?.let { put("countryIso", it) }
        }
        return withTimeoutOrNull(CALL_TIMEOUT_MS) {
            runCatching {
                FirebaseFunctions.getInstance(FUNCTIONS_REGION)
                    .getHttpsCallable(CALLABLE_NAME)
                    .call(payload)
                    .await()
                Analytics.sosCloudReported(status)
                Log.d(TAG, "SOS cloud report ok status=$status uid=${uid.takeLast(6)}")
                true
            }.onFailure { throwable ->
                Log.w(TAG, "SOS cloud report failed status=$status", throwable)
            }.getOrDefault(false)
        } ?: false
    }

    /** Restores the persisted Firebase session or signs in anonymously, time-bounded. */
    private suspend fun ensureSignedInUid(): String? {
        val auth = FirebaseAuth.getInstance()
        auth.currentUser?.uid?.let { return it }
        return withTimeoutOrNull(SIGN_IN_TIMEOUT_MS) {
            runCatching { auth.signInAnonymously().await().user?.uid }
                .onFailure { Log.w(TAG, "Anonymous sign-in for SOS report failed", it) }
                .getOrNull()
        }
    }

    /**
     * Best-effort ISO country for agency routing: telephony network country, then SIM country,
     * then reverse-geocoding the fix. Any failure just omits the hint (routes to International).
     */
    private suspend fun resolveCountryIso(
        context: Context,
        location: BlePeerIdentityUtils.PeerLocationSnapshot?
    ): String? {
        val telephony = context.getSystemService(Context.TELEPHONY_SERVICE) as? TelephonyManager
        val fromTelephony = runCatching {
            telephony?.networkCountryIso?.takeIf { it.isNotBlank() }
                ?: telephony?.simCountryIso?.takeIf { it.isNotBlank() }
        }.getOrNull()
        if (!fromTelephony.isNullOrBlank()) {
            return fromTelephony.uppercase(Locale.US)
        }
        if (location == null || !Geocoder.isPresent()) {
            return null
        }
        return withContext(Dispatchers.IO) {
            runCatching {
                @Suppress("DEPRECATION")
                Geocoder(context, Locale.US)
                    .getFromLocation(location.latitude, location.longitude, 1)
                    ?.firstOrNull()
                    ?.countryCode
                    ?.takeIf { it.isNotBlank() }
                    ?.uppercase(Locale.US)
            }.getOrNull()
        }
    }

    /**
     * Human-readable place of the fix (district/city) for the status UI. Cached until the victim
     * moves ~1 km so heartbeats don't re-geocode the same spot.
     */
    private var cachedPlace: Triple<Double, Double, String>? = null

    private suspend fun resolvePlaceName(
        context: Context,
        location: BlePeerIdentityUtils.PeerLocationSnapshot?
    ): String? {
        if (location == null || !Geocoder.isPresent()) {
            return cachedPlace?.third
        }
        cachedPlace?.let { (lat, lon, place) ->
            if (isWithinPlaceCacheRadius(lat, lon, location.latitude, location.longitude)) {
                return place
            }
        }
        val place = withContext(Dispatchers.IO) {
            runCatching {
                @Suppress("DEPRECATION")
                val address = Geocoder(context)
                    .getFromLocation(location.latitude, location.longitude, 1)
                    ?.firstOrNull()
                    ?: return@runCatching null
                listOfNotNull(
                    address.subLocality?.takeIf { it.isNotBlank() }
                        ?: address.locality?.takeIf { it.isNotBlank() }
                        ?: address.subAdminArea?.takeIf { it.isNotBlank() },
                    address.adminArea?.takeIf { it.isNotBlank() }
                )
                    .distinct()
                    .joinToString(", ")
                    .takeIf { it.isNotBlank() }
            }.getOrNull()
        }
        if (place != null) {
            cachedPlace = Triple(location.latitude, location.longitude, place)
        }
        return place ?: cachedPlace?.third
    }

    private fun isWithinPlaceCacheRadius(
        lat1: Double,
        lon1: Double,
        lat2: Double,
        lon2: Double
    ): Boolean {
        val results = FloatArray(1)
        return runCatching {
            android.location.Location.distanceBetween(lat1, lon1, lat2, lon2, results)
            results[0] <= PLACE_CACHE_RADIUS_METERS
        }.getOrDefault(false)
    }

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private fun Context.hasUsableInternet(): Boolean {
        val manager = getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return false
        val capabilities = manager.getNetworkCapabilities(manager.activeNetwork) ?: return false
        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
            capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
    }

    private const val STATUS_ACTIVE = "active"
    private const val STATUS_RESOLVED = "resolved"
    private const val PLACE_CACHE_RADIUS_METERS = 1_000f
}
